docs/proposals/last-operation-description.md

```diff
@@ -0,0 +1,68 @@
+# Last Operation Description
+
+## Introduction
+As per the OpenService Broker API spec, the last operation response can contain
+a description as well as a status [last operation response](https://github.com/openservicebrokerapi/servicebroker/blob/v2.12/spec.md#body).
+This can provide useful information about what is happening during a broker action along with details of the overall progress.
+UI implementors could make valuable use of this information
+to provide a progress bar or just simply a log of actions to show progress.
+```
+60%: succesfully created realm in keycloak
+```
+
+## Problem Description
+Currently we are only making use of the status field: 
+[last operation in broker.go](https://github.com/openshift/ansible-service-broker/blob/master/pkg/broker/broker.go#L1311)
+The difficulty is around how we get that detail back from the pod.
+Without the description, we limit the feedback that can be provided to the user.
+
+
+## Proposed Implementation
+Create an append only log file in a specified location ```/var/log/last_operation```, which is collected periodically and also a final time after
+the apb has completed but before the pod is deleted. 
+In the bind workflow, we already do something similar by execing into the container in order to collect
```

Comment:
I'm wondering if all of this is possible without using a file inside the APB container. Kubernetes already has a [downward API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/) for exposing things like Pod Name and Pod Namespace combined with [Kubernetes annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) means that we could simply have an annotation(s) for `last_operation` and `final_operation` available through API calls w/o needing to exec.

```
$ kubectl annotate pods asb-1-h597s last_operation='Last operation added'
pod "asb-1-h597s" annotated
$ kubectl get pods asb-1-h597s -o go-template='{{ .metadata.annotations.last_operation }}'
Last operation added%
```

Reply:
@djzager Interesting that you mention the annotations approach. This was something I also thought about. I shied away from it as the file approach seemed more fitting given the pod was already being exec'd into. I agree though this could be a simpler approach, particularly taking into account using the downward API to inform the apb. 
Curious on your thoughts about using a watch while the pod was executing, would this be a reasonable way to pick up updates to the annotations in the broker?
I will rework the proposal to make use of the downward API and pod annotations.

Reply:
So, the APB pod that is running would add annotations on itself?

Reply:
Yes that is my understanding. I believe the apb module could make this transparent to the apb developer

Reply:
I like the idea of using watch, unfortunately this is where I get a little fuzzy with how the broker handles the provision/deprovision jobs, their associated channels, and writing state information to etcd. My immediate impression is that we may need to make some changes to the broker in order to support additional status changes if we are going to write the last_operation to etcd.

You _may_ be able to get by with handling the `in progress` state with an associated request for the `last_operation` annotation on the pod when the service catalog is asking for the last operation.

Reply:
I am wondering if using the labels instead of the annotation might make more sense. I don't fully understand the difference but just starting the discussion.

Reply:
I am a bit hazy on the difference also.  However I know labels are used as selectors etc and for querying via the client. Whereas I have often seen annotations used for specific meta data for example https://github.com/openshift/origin/blob/master/examples/quickstarts/cakephp-mysql-persistent.json#L11
So I think that in this case annotations would make more sense.

Reply:
A question on this approach: If we go with the watch on the pod, I am wondering would it at some point also make sense to do the same thing for the credentials from a bind. Rather than execing into the pod, have the pod create a secret in the pods namespace and have the broker watch for secrets in that namespace. 
Obviously this is separate from this work but it could be a good paradigm to use for these kind of interactions? I can create a separate issue for this to discuss further.

Reply:
I don't know the security impact, if at all, of the apb creating a secret and populating it. But, I think it's a good idea.  The broker could also create it and the apb populate it.  There are a few avenues we could take here. We'll have to explore this separately.

> So I think that in this case annotations would make more sense.
>
+1 for annotations. I think labels are meant for kubernetes services identifying other kubernetes services.

> Curious on your thoughts about using a watch while the pod was executing, would this be a 
> reasonable way to pick up updates to the annotations in the broker?
>
If the broker watches the resource,  it should allow us to always get the final operation because the logic will exist in the broker.  We can grab the final operation right before the namespace is cleaned up. +1